pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/TableRebalancer.java

```diff
@@ -257,47 +390,162 @@ private void setTargetState(IdealState idealState, String segmentId, Map<String,
+            LOGGER.debug("Host-state mapping of segment {} has state {} in external view and state {} in ideal state",
+                segment, stateInExternalView, stateInIdealState);
+            // external view not yet converged
+            stable = false;
+          }
         }
       }
     }
-    return numDiff;
+    return stable;
+  }
+
+  private static void prettyPrintMapDebug(final Map<String, String> map) {
+    final Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator(":");
+    LOGGER.debug(mapJoiner.join(map));
+  }
+
+  private static void prettyPrintMapError(final Map<String, String> map) {
+    final Joiner.MapJoiner mapJoiner = Joiner.on(",").withKeyValueSeparator(":");
+    LOGGER.error(mapJoiner.join(map));
   }
 
   /**
    * Wait till state has stabilized {@link #isStable(String)}
    */
-  private void waitForStable(String resourceName)
-      throws InterruptedException {
-    int diff;
+  private void waitForStable(final String resourceName) {
     int INITIAL_WAIT_MS = 3000;
-    Thread.sleep(INITIAL_WAIT_MS);
-    do {
-      diff = isStable(resourceName);
-      if (diff == 0) {
-        break;
-      } else {
-        LOGGER.info(
-            "Waiting for externalView to match idealstate for table:" + resourceName + " Num segments difference:"
-                + diff);
-        Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
-      }
-    } while (diff > 0);
+    try {
+      Thread.sleep(INITIAL_WAIT_MS);
+      while (true) {
+        if (isStable(resourceName)) {
+          break;
+        } else {
+          LOGGER.info("Waiting for externalView to match idealstate for table:" + resourceName);
+          Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
+          }
+        }
+      } catch (InterruptedException e) {
+      LOGGER.error("Rebalancer got interrupted while waiting for external view to converge");
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  /**
+   * Helper class that maintains stats that
+   * are later checked in tests to verify
+   * the behavior of the algorithm here
+   * that takes from current ideal state
+   * to a target ideal state
+   */
+  public static class RebalancerStats {
+    private int dryRun;
+    private int updatestoIdealStateInZK;
+    private int directUpdatesToSegmentInstanceMap;
+    private int incrementalUpdatesToSegmentInstanceMap;
+
+    RebalancerStats() {
+    }
+
+    /**
+     * Number of dry runs. Can only be 1
+     * @return
+     */
+    public int getDryRun() {
+      return dryRun;
+    }
+
+    /**
+     * Get the number of times rebalancer
+     * successfully updated the ideal
+     * state in ZK as it progresses through
+     * rebalancing.
+     * @return ideal state updates in ZK
+     */
+    public int getUpdatestoIdealStateInZK() {
+      return updatestoIdealStateInZK;
+    }
+
+    /**
+     * Get the number of times we updated the instance-state
```

Comment:
I am not clear on what this means. Is it the max number of times we updated in one step?

Reply:
No, this is the total number of time we updated the host-state mapping of a segment to target mapping in a single step. This is cumulative across all segments.